gnucash/import-export/import-backend.c

```diff
@@ -827,6 +839,90 @@ void split_find_match (GNCImportTransInfo * trans_info,
 /***********************************************************************
  */
 
+/* append the imported transaction description to the matched transaction description */
+static void
+desc_append (const GNCImportTransInfo* trans_info)
+{
+    GNCImportMatchInfo *selected_match =
+                gnc_import_TransInfo_get_selected_match (trans_info);
+
+    gchar* tmp = g_strconcat( xaccTransGetDescription (selected_match->trans),
+                              "|",
+                              xaccTransGetDescription (gnc_import_TransInfo_get_trans( trans_info)),
+                              NULL);
+    xaccTransSetDescription (selected_match->trans, tmp);
+    g_free (tmp);
+}
+
+/* replace the matched transaction description with the imported transaction description */
+static void
+desc_replace (const GNCImportTransInfo* trans_info)
+{
+    GNCImportMatchInfo *selected_match =
+                gnc_import_TransInfo_get_selected_match (trans_info);
+
+    xaccTransSetDescription (selected_match->trans,
+                             xaccTransGetDescription (
+                                gnc_import_TransInfo_get_trans (trans_info)));
+}
+
+/* append the imported transaction notes to the matched transaction notes */
+static void
+notes_append (const GNCImportTransInfo* trans_info)
+{
+    GNCImportMatchInfo *selected_match =
+                gnc_import_TransInfo_get_selected_match (trans_info);
+
+    gchar* tmp = g_strconcat (xaccTransGetNotes (selected_match->trans),
+                              "|",
+                              xaccTransGetNotes (gnc_import_TransInfo_get_trans (trans_info)),
+                              NULL);
+    xaccTransSetNotes (selected_match->trans, tmp);
+    g_free (tmp);
+}
+
+/* replace the matched transaction notes with the imported transaction notes */
+static void
+notes_replace (const GNCImportTransInfo* trans_info)
+{
+    GNCImportMatchInfo *selected_match =
+                gnc_import_TransInfo_get_selected_match(trans_info);
+
+    xaccTransSetNotes(selected_match->trans,
+                      xaccTransGetNotes(
+                          gnc_import_TransInfo_get_trans(trans_info)));
+}
+
+/* Append or replace transaction description and notes
+ * depending on the Append checkbox
+ */
+static void
+update_desc_and_notes (const GNCImportTransInfo* trans_info)
+{
+    GNCImportMatchInfo* selected_match =
+            gnc_import_TransInfo_get_selected_match (trans_info);
+
+    if (trans_info->append_text)
+    {
+        // Append if imported desc not already in matched desc
+        if ( !g_strrstr (xaccTransGetDescription ( selected_match->trans),
```

Comment:
Why `g_strrstr` instead of `strstr`? Also you should call `g_utf8_normalize()` on both strings before searching to ensure that the representations of complex characters are the same.
It might also be worthwhile to check that `g_utf8_strlen(selected_match->trans) >= g_utf8_strlen(trans_info)` before doing a potentially expensive substring search.

Reply:
@jralls,
Re g_strrstr: Totally agree, strstr would be better. I cannot remember how I decided on g_strrstr. I shall fix.

Reply:
Oops, last comment was not finished because I keep forgetting not to do Ctrl-Enter for a newline, as in another app I often use, as that is equivalent to clicking 'comment' button in github.

Re normalize: Good idea, I shall do.

Re checking g_utf_strlen:
Please correct me, but it seems to me parsing 2 strings into utf-8 characters in order to count them would be much more processing than doing the strstr() byte by byte comparison?

Reply:
> Please correct me, but it seems to me parsing 2 strings into utf-8 characters in order to count them would be much more processing than doing the strstr() byte by byte comparison?

Well, if you normalize first then you can compare the `strlen`s instead, that would save having to decode the utf8 sequences into gunichars twice (once for the nomalization and again for the length calculation). Even faster would be to convert both strings to normalized UTF32 (i.e. `gunichar`), normalize, then count and search the UTF32 because it would reduce the number of comparisons in languages other than English. Unfortunately GLib doesn't provide that functionality so you'd have to use ICU. That extra complication isn't worth the effort.

`strlen` is faster than char-by-char comparisons of two strings because it needs only one register load per character and has a very simple branching structure. `strstr` OTOH can be quite expensive because it may have to backtrack if there are partial matches of 'needle` in `haystack`.